File: backend/src/main/java/com/backend/ord/domain/entities/UserProgress.java
package com.backend.ord.domain.entities;

import com.backend.ord.domain.entities.abstracts.EntityBase;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

@Entity
@Table(name = "user_progress")
public class UserProgress extends EntityBase {
    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "game_id", nullable = false)
    private Game game;

    @NotNull
    @Column(name = "points_obtained", nullable = false)
    private Integer pointsObtained;

    public @NotNull User getUser() {
        return this.user;
    }

    public @NotNull Game getGame() {
        return this.game;
    }

    public @NotNull Integer getPointsObtained() {
        return this.pointsObtained;
    }

    public void setUser(@NotNull User user) {
        this.user = user;
    }

    public void setGame(@NotNull Game game) {
        this.game = game;
    }

    public void setPointsObtained(@NotNull Integer pointsObtained) {
        this.pointsObtained = pointsObtained;
    }
}

File: backend/src/main/kotlin/com/backend/ord/domain/entities/UserProgress.kt
package com.backend.ord.domain.entities

import jakarta.persistence.*
import jakarta.validation.constraints.Size
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.OnDelete
import org.hibernate.annotations.OnDeleteAction
import org.hibernate.annotations.UpdateTimestamp
import java.time.Instant
import java.util.*

@Entity
@Table(name = "user_progress")
class UserProgress(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    var id: UUID = UUID.randomUUID(),

    @Column(name = "points_obtained", nullable = false)
    @field:Size(min = 0)
    var pointsObtained: Int = 0,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "user_id", nullable = false)
    var user: User,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "game_id", nullable = false)
    var game: Game,

    @Column(name = "created_at", nullable = false, updatable = false)
    @CreationTimestamp
    var createdAt: Instant = Instant.now(),

    @Column(name = "updated_at", nullable = false)
    @UpdateTimestamp
    var updatedAt: Instant = Instant.now()
)